### PR TITLE
Feat liquidations

### PR DESCRIPTION
**Pull request by poliwop**

### CODE REVIEW

File: hydradx/tests/test_otc.py [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+    if otc.sell_amount != init_sell_amt - sell_amt:
+        raise
+    # DOT conserved
+    if (omnipool.liquidity['DOT'] + otc.buy_amount + treasury.holdings['DOT'] != pytest.approx(
+            init_state.liquidity['DOT'], rel=1e-15)):
+        raise
+    # USDT conserved
+    if omnipool.liquidity['USDT'] + treasury.holdings['USDT'] != pytest.approx(init_state.liquidity['USDT'] + init_sell_amt, rel=1e-15):
+        raise
+    if treasury.holdings['DOT'] <= 0:  # treasury should make profit
```
Comment (jepidoptera) [inline hydradx/tests/test_otc.py:120]: I think that correct practice is to put this text (treasury should make profit) into an error message, not a comment.

File: hydradx/tests/test_otc.py [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+
+    # OTC partially fillable
+    otc = OTC('DOT', 'USDT', sell_amount, price, partially_fillable=True)
+    partial_otc_amount = find_partial_otc_sell_amount(init_state, otc)
+    omnipool = init_state.copy()
+    treasury = Agent(holdings={"USDT": 0, "DOT": 0})
+    omnipool_settle_otc(omnipool, otc, treasury, partial_otc_amount)
+    spot_after = omnipool.buy_spot('DOT', 'USDT')
+
+    otc.validate()
```
Comment (jepidoptera) [inline hydradx/tests/test_otc.py:271]: This doesn't do anything.

**Review by jepidoptera — CHANGES_REQUESTED** [review]

File: hydradx/model/amm/otc.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    def __repr__(self):
+        return f"OTC({self.buy_asset}, {self.sell_asset}, {self.sell_amount}, {self.price}, {self.partially_fillable})"
+
+    def copy(self):
+        return OTC(self.buy_asset, self.sell_asset, self.sell_amount, self.price, self.partially_fillable)
+
+    # numeraire is asset OTC order is selling
+    def sell(self, agent: Agent, sell_amount: float) -> None:
+        '''sell_amount is the amount of the buy_asset that the agent is selling for the sell_asset.'''
```
[the remark this answers is not in the document]
Comment (poliwop) [reply hydradx/model/amm/otc.py:21]: Pycharm is wrong IMO

File: hydradx/model/amm/global_state.py [hunk 4]
```diff
@@ -358,3 +370,170 @@ def transform(state: GlobalState) -> GlobalState:
...
+            if cdp.in_liquidation:
+                delta_debt = find_partial_liquidation_amount(omnipool, cdp, penalty)
+                if delta_debt > 0:
+                    omnipool_liquidate_cdp(state.pools['omnipool'], cdp, agent, delta_debt, state.liquidation_penalty)
+        return state
+
+    return transform
+
+
+def find_partial_liquidation_amount(omnipool: OmnipoolState, cdp: CDP, penalty: float, iters: float = 20, min_amt = 1) -> float:
```
[the remark this answers is not in the document]
Comment (poliwop) [reply hydradx/model/amm/global_state.py:390]: This is true but I believe the reason for this parameter is rounding error. A parameter for setting minimum value would be implemented differently.
I think I actually implemented this chasing a rounding error bug that turned out to be something else. For now I will try simply setting default value to 0.

File: hydradx/model/amm/global_state.py [hunk 5]
```diff
@@ -358,3 +370,170 @@ def transform(state: GlobalState) -> GlobalState:
...
+                break
+        delta_debt = (delta_debt_up + delta_debt_down) / 2
+    return delta_debt_down
+
+
+def omnipool_liquidate_cdp(omnipool: OmnipoolState, cdp: CDP, treasury_agent: Agent, delta_debt: float, penalty: float) -> None:
+    # treasury_agent buys borrowed asset to cover the debt, but only if debt can be covered by existing collateral
+    # check if debt can be covered by existing collateral
+    if delta_debt > cdp.debt_amt or delta_debt == 0:
+        return
```
[the remark this answers is not in the document]
Comment (poliwop) [reply hydradx/model/amm/global_state.py:436]: I guess for now an exception is okay if delta_debt > cdp.debt_amt. I think if delta_debt == 0, that is valid and nothing should happen.

File: hydradx/model/amm/global_state.py [hunk 6]
```diff
@@ -358,3 +370,170 @@ def transform(state: GlobalState) -> GlobalState:
...
+    if cdp.collateral_asset not in treasury_agent.holdings:
+        treasury_agent.holdings[cdp.collateral_asset] = 0
+    init_collat_amt = cdp.collateral_amt
+    treasury_agent.holdings[cdp.collateral_asset] += init_collat_amt  # flash mint collateral_amt to treasury_agent
+    omnipool_check, agent_check = simulate_swap(
```
[the remark this answers is not in the document]
Comment (poliwop) [reply hydradx/model/amm/global_state.py:441]: Had to add a case that calculate_sell_from_buy did not handle correctly, but with that change I was also able to make the change you suggest.

**Review by jepidoptera — APPROVED** [review]

**Pull request merged**